File: Data Structures & Algorithms/string-encode-and-decode/submission-3.py
class Solution:

    def encode(self, strs: List[str]) -> str:
        if len(strs) == 0:
            return ""
        res = ""
        for s in strs:
            res += str(len(s))
            res += "#"
            res += s
        return res

    def decode(self, s: str) -> List[str]:
        if s == "":
            return []
        
        i = 0
        res = []
        while i < len(s):
            j = i
            while s[j] != "#":
                j += 1
            length = int(s[i:j])
            res.append(s[j+1:j+length+1])
            i = j + length + 1
        return res

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-3.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        auto comp = [](ListNode* a, ListNode* b){
            return a->val > b->val;
        };
        priority_queue<ListNode*, vector<ListNode*>, decltype(comp)> minHeap(comp);
        ListNode* res = new ListNode(0);
        ListNode* dummy = res;
        for(auto l : lists){
            if(l) minHeap.push(l);
        }

        while(!minHeap.empty()){
            ListNode* tmp = minHeap.top();
            minHeap.pop();
            if(tmp->next) minHeap.push(tmp->next);

            dummy->next = tmp;
            dummy = dummy->next;
        }
        return res->next;
    }
};
